stats/stats-server/src/update_service.rs

```diff
@@ -49,32 +55,45 @@ impl UpdateService {
                 let group_entry = group.clone();
                 let default_schedule = default_schedule.clone();
                 let sema = semaphore.clone();
-                async move {
-                    let _permit = sema.acquire().await.expect("failed to acquire permit");
-                    if let Some(force_full) = force_update_on_start {
-                        this.clone().update(group_entry.clone(), force_full).await
-                    };
-                    this.spawn_group_updater(group_entry, &default_schedule);
-                }
+                (
+                    async move {
+                        let _permit = sema.acquire().await.expect("failed to acquire permit");
+                        if let Some(force_full) = force_update_on_start {
+                            this.clone().update(group_entry.clone(), force_full).await
+                        };
+                    },
+                    self.spawn_group_updater(group.clone(), &default_schedule),
+                )
             })
-            .collect::<Vec<_>>();
+            .collect::<(Vec<_>, Vec<_>)>();
         futures::future::join_all(tasks).await;
         tracing::info!("initial update is done");
+
+        let mut failed = 0;
+        while !updaters.is_empty() {
+            let (res, _, others) = futures::future::select_all(updaters).await;
+            updaters = others;
+            tracing::error!("updater stopped: {:?}", res);
+            failed += 1;
+            if failed >= FAILED_UPDATERS_UNTIL_PANIC {
+                panic!("too many failed updaters");
+            }
+        }
     }
 
     fn spawn_group_updater(
         self: &Arc<Self>,
         group_entry: UpdateGroupEntry,
         default_schedule: &Schedule,
-    ) {
+    ) -> JoinHandle<()> {
         let this = self.clone();
         let chart = group_entry.clone();
         let schedule = group_entry
             .update_schedule
             .as_ref()
             .unwrap_or(default_schedule)
             .clone();
-        tokio::spawn(async move { this.run_cron(chart, schedule).await });
+        tokio::spawn(async move { this.run_cron(chart, schedule).await })
```

Comment:
Nit: probably should do the same
```suggestion
        tokio::spawn(this.run_cron(chart, schedule))
```